docs/adapting-settings.md
# Adapting settings

You can specify a configuration file by exporting the `UDATA_SETTINGS` environment variable.

```shell
export UDATA_SETTINGS=/path/to/my/udata.cfg
```

The configuration file is simply a Python file declaring variables.

udata uses a few Flask extensions and therefore provides all available options for these extensions.
Most of the time, it tries to provide sane defaults.

## Flask and global behavior options

### DEBUG

**default**: `False`

A boolean specifying the debug mode.

### SEND_MAIL

**default**: `True`

A boolean specifying if the emails should actually be sent.

### DEFAULT_LANGUAGE

**default**: `'en'`

The default fallback language when no language prefix is provided in URLs.

### SECRET_KEY

A secret key used as salt for cryptographic parts.
**You must specify your own secure key and use the same in all your instances**.

### SITE_ID

**default**: `'default'`

The site identifier. It is used to attached some database configuration, metrics...

### SITE_TERMS_LOCATION

**default**: `generic embedded terms`

The site terms in markdown. It can be either an URL or a local path to a markdown content.
If this is an URL, the content is downloaded on the first terms page display and cached.

### PLUGINS

**default**: `[]`

A list of enabled udata plugins.

### THEME

**default**: ``'default'``

The enabled theme name.

### TEMPLATE_CACHE_DURATION

**default**: `5`

The duration used for templates' cache, in minutes.

### ALLOWED_RESOURCES_EXTENSIONS

**default**:
```python
[
    # Base
    'csv', 'txt', 'json', 'pdf', 'xml', 'rdf', 'rtf', 'xsd',
    # OpenOffice
    'ods', 'odt', 'odp', 'odg',
    # Microsoft Office
    'xls', 'xlsx', 'doc', 'docx', 'pps', 'ppt',
    # Archives
    'tar', 'gz', 'tgz', 'rar', 'zip', '7z', 'xz', 'bz2',
    # Images
    'jpeg', 'jpg', 'jpe', 'gif', 'png', 'dwg', 'svg', 'tiff', 'ecw', 'svgz', 'jp2',
    # Geo
    'shp', 'kml', 'kmz', 'gpx', 'shx', 'ovr', 'geojson',
    # Meteorology
    'grib2',
    # Misc
    'dbf', 'prj', 'sql', 'epub', 'sbn', 'sbx', 'cpg', 'lyr', 'owl',
]
```

This is the allowed resources extensions list that user can upload.

### RESOURCES_FILE_ALLOWED_DOMAINS

**default**: `[]`

Whitelist of urls domains allowed for resources with `filetype` equals to `file`.

`SERVER_NAME` is always included.

`*` is a supported value as a wildcard allowing all domains.

### PREVIEW_MODE

**default**: `'iframe'`

Define the resources preview mode. Can be one of:
- `'iframe'`: preview are displayed into an iframe modal
- `'page'`: preview is displayed into a new page

If you want to disable preview, set `PREVIEW_MODE` to `None`

### ARCHIVE_COMMENT_USER_ID

**default**: `None`

The id of an existing user which will post a comment when a dataset is archived.

### ARCHIVE_COMMENT_TITLE

**default**: `_('This dataset has been archived')`

The title of the comment optionaly posted when a dataset is archived.
NB: the content of the comment is located in `udata/templates/comments/dataset_archived.txt`.


### SCHEMA_CATALOG_URL

**default** : `None`

The URL to a schema catalog, listing schemas resources can conform to. The URL should be a JSON endpoint, returning a schema catalog. Example: https://schema.opendata.gouv.tg /schemas/schemas.json

NB: this is used by the `datasets/schemas` API to fill the `schema` field of a `Resource`.

## URLs validation

### URLS_ALLOW_PRIVATE

**default**:  `False`

Whether or not to allow private URLs (private IPs...) submission

### URLS_ALLOW_LOCAL

**default**: `False`

Whether or not to allow local URLs (localhost...) submission.
When developping you might need to set this to `True`.

### URLS_ALLOW_CREDENTIALS

**default**: `True`

Whether or not to allow credentials in URLs submission.

### URLS_ALLOWED_SCHEMES

**default**: `('http', 'https', 'ftp', 'ftps')`

List of allowed URL schemes.

### URLS_ALLOWED_TLDS

**default**: All IANA registered TLDs

List of allowed TLDs.
When using udata on an intranet, you might want to add your own custom TLDs:

```python
from udata.settings import Defaults

URLS_ALLOWED_TLDS = Defaults.URLS_ALLOWED_TLDS + set(['custom', 'company'])
```

### EXPORT_CSV_MODELS

**default**: `('dataset', 'resource', 'discussion', 'organization', 'reuse', 'tag')`

List models that will be exported to CSV by the job `export-csv`.
You can disable the feature by setting this to an empty list.

### EXPORT_CSV_DATASET_ID

**default**: `None`

The id of a dataset that should be created before running the `export-csv` job and will hold the CSV exports.

## Search configuration

### SEARCH_AUTOCOMPLETE_ENABLED

**default**: `True`

Enables the search autocomplete on frontend if set to `True`, disables otherwise.

### SEARCH_AUTOCOMPLETE_DEBOUNCE

**default**: `200`

### SEARCH_SERVICE_API_URL

**default**: None

The independent search service api url to use if available.
If not specified, mongo full text search is used.

Ex:
```python
SEARCH_SERVICE_API_URL = 'http://127.0.0.1:5000/api/1/'
```

See [udata-search-service][udata-search-service] for more information on using a search service.

## Spatial configuration

### SPATIAL_SEARCH_EXCLUDE_LEVELS

**default**: `tuple()`

List spatial levels that shoudn't be indexed (for time, performance and user experience).

## Territories configuration

### ACTIVATE_TERRITORIES

**default**: `False`

Whether you want to activate pages and API related to territories.
Don't forget to set the `HANDLED_LEVELS` setting too.

### HANDLED_LEVELS

**default**: `tuple()`

The list of levels that you want to deal with.

Warning: the order is important and will determine parents/children for
a given territory. You have to set the smallest territory level first:

```python
HANDLED_LEVELS = ('fr:commune', 'fr:departement', 'fr:region')
```

## Harvesting configuration

### HARVEST_PREVIEW_MAX_ITEMS

**default**: `20`

The number of items to fetch while previewing an harvest source

### HARVEST_MAX_ITEMS

**default**: `None`

The max number of items to fetch when harvesting (development setting)

### HARVEST_DEFAULT_SCHEDULE

**default**: `0 0 * * *`

A cron expression used as default harvester schedule when validating harvesters.

### HARVEST_JOBS_RETENTION_DAYS

**default**: `365`

The number of days of harvest jobs to keep (ie. number of days of history kept)

## Link checker configuration

### LINKCHECKING_ENABLED

**default**: `True`

A flag to enable the resources urls check by an external link checker.

### LINKCHECKING_DEFAULT_LINKCHECKER

**default**: `no_check`

An entrypoint key of `udata.linkcheckers` that will be used as a default link checker, i.e. when no specific link checker is set for a resource (via `resource.extras.check:checker`).

### LINKCHECKING_IGNORE_DOMAINS

**default**: []

A list of domains to ignore when triggering link checking of resources urls.

### LINKCHECKING_IGNORE_PATTERNS

**default**: ['format=shp']

A list patterns found in checked URL to ignore (ie `pattern in url`).

### LINKCHECKING_MIN_CACHE_DURATION

**default**: 60

The minimum time in minutes between two consecutive checks of a resource's url.

### LINKCHECKING_MAX_CACHE_DURATION

**default**: 1080

The maximum time in minutes between two consecutive checks of a resource's url.

### LINKCHECKING_UNAVAILABLE_THRESHOLD

**default**: 100

The number of unavailable checks after which the resource is considered lastingly unavailable and won't be checked as often.

## Mongoengine/Flask-Mongoengine options

### MONGODB_HOST

**default**: `'mongodb://localhost:27017/udata'`

The mongodb database used by udata.
During tests, the test database will use the same name suffixed by `-test`

See [the official Flask-MongoEngine documentation][flask-mongoengine-doc]
for more details.

Authentication is also supported in the URL:

```python
MONGODB_HOST = 'mongodb://<user>:<password>@<host>:<port>/<database>'
```

### MONGODB_HOST_TEST

**default**: same as `MONGODB_HOST` with a `-test` suffix on the collection

An optional alternative mongo database used for testing.

## Celery options

By default, udata is configured to use Redis as Celery backend and a customized MongoDB scheduler.

The defaults are:

```python
CELERY_BROKER_URL = 'redis://localhost:6379'
CELERY_BROKER_TRANSPORT_OPTIONS = {
    'fanout_prefix': True,
    'fanout_patterns': True,
}
CELERY_RESULT_BACKEND = 'redis://localhost:6379'
CELERY_ACCEPT_CONTENT = ['pickle', 'json']
CELERY_WORKER_HIJACK_ROOT_LOGGER = False
CELERY_BEAT_SCHEDULER = 'udata.tasks.Scheduler'
CELERY_MONGODB_SCHEDULER_COLLECTION = "schedules"
```

Authentication is supported on Redis:

```python
CELERY_RESULT_BACKEND = 'redis://u:<password>@<host>:<port>'
CELERY_BROKER_URL = 'redis://u:<password>@<host>:<port>'
```

You can see the full list of Celery options in the [Celery official documentation][celery-doc].

**Note** Celery parameters changed in udata 1.2 because Celery has been upgraded to 4.1.0.
(You can get the change map [here][celery-conf-map]).
udata expect Celery parameters to be upper case and prefixed by `CELERY_` in your `udata.cfg`
and they will be automatically transformed for Celery 4.x:
Example:
 - Celery 3.x expected `BROKER_URL` and Celery 4.x expects `broker_url` so you need to change `BROKER_URL` to `CELERY_BROKER_URL` in your settings
 - Celery 3.X expected `CELERY_RESULT_BACKEND` and Celery 4.x expects `result_backend` so you can leave `CELERY_RESULT_BACKEND`

## Flask-Mail options

You can see the full configuration option list in
[the official Flask-Mail documentation][flask-mail-doc].

### MAIL_DEFAULT_SENDER

**default**: `'webmaster@udata'`

The default identity used for outgoing mails.

## Authlib options

udata uses Authlib to provide OAuth2 on the API.
The full option list is available in

CHANGELOG.md
# Changelog

## Current (in progress)

- Add more comments and types in the `api_field.py` "lib" [#3174](https://github.com/opendatateam/udata/pull/3174)

## 10.0.0 (2024-11-07)

- **breaking change** Migrate organization badges label to lowercase [#3182](https://github.com/opendatateam/udata/pull/3182)
- Add contact_form in ContactPoint api fields [#3175](https://github.com/opendatateam/udata/pull/3175)
- Expose contact point in DCAT RDF [#3179](https://github.com/opendatateam/udata/pull/3179)
- Use trailing slashes for the upload files URLs [#3177](https://github.com/opendatateam/udata/pull/3177)
- Use hydra's RESTful endpoint URLs [#3146](https://github.com/opendatateam/udata/pull/3146)
- Return dataservices visible to the user [#3180](https://github.com/opendatateam/udata/pull/3180)
- Fix flaky "duplicated email" importing fixtures tests [#3176](https://github.com/opendatateam/udata/pull/3176)
- Fix deprecated CircleCI config [#3181](https://github.com/opendatateam/udata/pull/3181)
- Use proper RESTful Hydra API endpoints [#3178](https://github.com/opendatateam/udata/pull/3178)
- Add a "filter by organization badge" for datasets, dataservices, reuses and organizations [#3155](https://github.com/opendatateam/udata/pull/3155)
    * you will need https://github.com/opendatateam/udata-search-service/pull/49 for the search service
- Add dataservices search with search-service [#3171](https://github.com/opendatateam/udata/pull/3171)
    * you will need https://github.com/opendatateam/udata-search-service/pull/48
- Expose the "landingPage" in DCAT RDF [#3183](https://github.com/opendatateam/udata/pull/3183)
- Licence.guess: extract first URL for better matching [#3185](https://github.com/opendatateam/udata/pull/3185)
- Expose the dataset's alternate identifier in RDF [#3186](https://github.com/opendatateam/udata/pull/3186)

## 9.2.4 (2024-10-22)

- Add a job to bind Tabular API to its datasets [#3162](https://github.com/opendatateam/udata/pull/3162)
- Expose dataservices' datasets by link instead of list [#3156](https://github.com/opendatateam/udata/pull/3156)
- CSW ISO Harvest: add extra_configs.remote_url_prefix [#3157](https://github.com/opendatateam/udata/pull/3157)
- DCAT harvest : Refactor DCAT extras handling in rdf parsing [#3054](https://github.com/opendatateam/udata/pull/3054). `accesRights`, `rights` and `license` are now in `extras["dcat"]` for both datasets and resources. `accessRights` can be infered for a dataset from its resources.

## 9.2.3 (2024-10-14)

- Add migration to delete duplicate resources due to ODS harvesting [#3158](https://github.com/opendatateam/udata/pull/3158)
- Add discussion.posted_on in discussion sort choices [3168](https://github.com/opendatateam/udata/pull/3168)

## 9.2.2 (2024-10-08)

- Add a filter on organization and document sort parameters in the `/discussions` endpoint [#3147](https://github.com/opendatateam/udata/pull/3147)
- Move discussion catalog creation and add fields [#3152](https://github.com/opendatateam/udata/pull/3152) and [#3154](https://github.com/opendatateam/udata/pull/3154)
- Add resources formats and harvest remote_url on dataset catalog [#3159](https://github.com/opendatateam/udata/pull/3159)
- Add contact form in contact point model [#3164](https://github.com/opendatateam/udata/pull/3164)
- Make base_api_url optional in dataservice [https://github.com/opendatateam/udata/pull/3163](#3163)

## 9.2.1 (2024-09-23)

- Enable basic search on dataservices [#3148](https://github.com/opendatateam/udata/pull/3148)

## 9.2.0 (2024-09-13)

- Allow OAuth clients without secrets [#3138](https://github.com/opendatateam/udata/pull/3138)
- Add a `archived` button for datasets and reuses on frontend admin [#3104](https://github.com/opendatateam/udata/pull/3104)
- **breaking change** Return all the reuses available to a user on the /reuses endpoint, including the private and deleted ones they own [#3140](https://github.com/opendatateam/udata/pull/3140).
- Fix undelete reuse and dataservices [#3141](https://github.com/opendatateam/udata/pull/3141)
- Add a minimal publiccode.yml [#3144](https://github.com/opendatateam/udata/pull/3144)
- Fix the boolean filters in the API for the "new system" endpoints [#3139](https://github.com/opendatateam/udata/pull/3139)
- Update authlib dependency from 0.14.3 to 1.3.1 [#3135](https://github.com/opendatateam/udata/pull/3135)
- Add CORS on resource redirect [#3145](https://github.com/opendatateam/udata/pull/3145)

## 9.1.4 (2024-08-26)

- Fix many linting issues reported by ruff [#3118](https://github.com/opendatateam/udata/pull/3118)
- Import the dataservice's organization from the fixtures [#3121](https://github.com/opendatateam/udata/pull/3121)
- Convert reuse to new API system [#3066](https://github.com/opendatateam/udata/pull/3066)

udata/settings.py
import pkg_resources
from kombu import Exchange, Queue
from tlds import tld_set

from udata.i18n import lazy_gettext as _

HOUR = 60 * 60


class Defaults(object):
    DEBUG = False
    TESTING = False
    SEND_MAIL = True
    LANGUAGES = {
        "en": "English",
        "fr": "Français",
        # "es": "Español",
        # "pt": "Português",
        # "sr": "Српски",
        # "de": "Deutsch",
    }
    DEFAULT_LANGUAGE = "fr"
    SECRET_KEY = "Default uData secret key"
    CONTACT_EMAIL = "contact@example.org"
    TERRITORIES_EMAIL = "territories@example.org"

    MONGODB_HOST = "mongodb://localhost:27017/udata"
    MONGODB_CONNECT = False  # Lazy connexion for Fork-safe usage

    # Search service configuration
    SEARCH_SERVICE_API_URL = None
    SEARCH_SERVICE_REQUEST_TIMEOUT = 20

    # BROKER_TRANSPORT = 'redis'
    CELERY_BROKER_URL = "redis://localhost:6379"
    CELERY_BROKER_TRANSPORT_OPTIONS = {
        "fanout_prefix": True,
        "fanout_patterns": True,
    }
    CELERY_RESULT_BACKEND = "redis://localhost:6379"
    CELERY_RESULT_EXPIRES = 6 * HOUR  # Results are kept 6 hours
    CELERY_TASK_IGNORE_RESULT = True
    CELERY_TASK_SERIALIZER = "pickle"
    CELERY_RESULT_SERIALIZER = "pickle"
    CELERY_ACCEPT_CONTENT = ["pickle", "json"]
    CELERY_WORKER_HIJACK_ROOT_LOGGER = False
    CELERY_BEAT_SCHEDULER = "udata.tasks.Scheduler"
    CELERY_MONGODB_SCHEDULER_COLLECTION = "schedules"
    CELERY_MONGODB_SCHEDULER_CONNECTION_ALIAS = "udata_scheduler"

    # Default celery routing
    CELERY_TASK_DEFAULT_QUEUE = "default"
    CELERY_TASK_QUEUES = (
        # Default queue (on default exchange)
        Queue("default", routing_key="task.#"),
        # High priority for urgent tasks
        Queue("high", Exchange("high", type="topic"), routing_key="high.#"),
        # Low priority for slow tasks
        Queue("low", Exchange("low", type="topic"), routing_key="low.#"),
    )
    CELERY_TASK_DEFAULT_EXCHANGE = "default"
    CELERY_TASK_DEFAULT_EXCHANGE_TYPE = "topic"
    CELERY_TASK_DEFAULT_ROUTING_KEY = "task.default"
    CELERY_TASK_ROUTES = "udata.tasks.router"

    CACHE_KEY_PREFIX = "udata-cache"
    CACHE_TYPE = "flask_caching.backends.redis"

    # Flask mail settings

    MAIL_DEFAULT_SENDER = "webmaster@udata"

    # Flask security settings

    SESSION_COOKIE_SECURE = True
    SESSION_COOKIE_SAMESITE = None  # Can be set to 'Lax' or 'Strict'. See https://flask.palletsprojects.com/en/2.3.x/security/#security-cookie

    # Flask-Security-Too settings

    SECURITY_TRACKABLE = True
    SECURITY_REGISTERABLE = True
    SECURITY_CONFIRMABLE = True
    SECURITY_RECOVERABLE = True
    SECURITY_CHANGEABLE = True

    SECURITY_PASSWORD_HASH = "bcrypt"
    SECURITY_PASSWORD_NORMALIZE_FORM = "NFKD"
    SECURITY_PASSWORD_LENGTH_MIN = 8
    SECURITY_PASSWORD_REQUIREMENTS_LOWERCASE = True
    SECURITY_PASSWORD_REQUIREMENTS_DIGITS = True
    SECURITY_PASSWORD_REQUIREMENTS_UPPERCASE = True
    SECURITY_PASSWORD_REQUIREMENTS_SYMBOLS = False

    SECURITY_PASSWORD_SALT = "Default uData secret password salt"
    SECURITY_CONFIRM_SALT = "Default uData secret confirm salt"
    SECURITY_RESET_SALT = "Default uData secret reset salt"
    SECURITY_REMEMBER_SALT = "Default uData remember salt"

    SECURITY_EMAIL_SENDER = MAIL_DEFAULT_SENDER

    SECURITY_EMAIL_SUBJECT_REGISTER = _("Welcome")
    SECURITY_EMAIL_SUBJECT_CONFIRM = _("Please confirm your email")
    SECURITY_EMAIL_SUBJECT_PASSWORDLESS = _("Login instructions")
    SECURITY_EMAIL_SUBJECT_PASSWORD_NOTICE = _("Your password has been reset")
    SECURITY_EMAIL_SUBJECT_PASSWORD_CHANGE_NOTICE = _("Your password has been changed")
    SECURITY_EMAIL_SUBJECT_PASSWORD_RESET = _("Password reset instructions")

    SECURITY_RETURN_GENERIC_RESPONSES = False

    # Sentry configuration
    SENTRY_DSN = None
    SENTRY_TAGS = {}
    SENTRY_USER_ATTRS = ["slug", "email", "fullname"]
    SENTRY_LOGGING = "WARNING"
    SENTRY_IGNORE_EXCEPTIONS = []
    SENTRY_SAMPLE_RATE: float = 1.0

    # Flask WTF settings
    CSRF_SESSION_KEY = "Default uData csrf key"

    # Flask-Sitemap settings
    # TODO: chose between explicit or automagic for params-less endpoints
    # SITEMAP_INCLUDE_RULES_WITHOUT_PARAMS = False
    SITEMAP_BLUEPRINT_URL_PREFIX = None

    AUTO_INDEX = True

    SITE_ID = "default"
    SITE_TITLE = "opendata.gouv.tg "
    SITE_KEYWORDS = ["opendata", "udata"]
    SITE_AUTHOR_URL = None
    SITE_AUTHOR = "Udata"
    SITE_GITHUB_URL = "https://github.com/etalab/udata"
    SITE_TERMS_LOCATION = pkg_resources.resource_filename(__name__, "terms.md")

    UDATA_INSTANCE_NAME = "udata"

    PLUGINS = []
    THEME = None

    STATIC_DIRS = []

    # OAuth 2 settings
    OAUTH2_PROVIDER_ERROR_ENDPOINT = "oauth.oauth_error"
    OAUTH2_REFRESH_TOKEN_GENERATOR = True
    OAUTH2_TOKEN_EXPIRES_IN = {
        "authorization_code": 30 * 24 * HOUR,
        "implicit": 10 * 24 * HOUR,
        "password": 30 * 24 * HOUR,
        "client_credentials": 30 * 24 * HOUR,
    }
    OAUTH2_ALLOW_WILDCARD_IN_REDIRECT_URI = False

    MD_ALLOWED_TAGS = [
        "a",
        "abbr",
        "acronym",
        "b",
        "br",
        "blockquote",
        "code",
        "dd",
        "del",
        "div",
        "dl",
        "dt",
        "em",
        "h1",
        "h2",
        "h3",
        "h4",
        "h5",
        "h6",
        "hr",
        "i",
        "img",
        "li",
        "ol",
        "p",
        "pre",
        "small",
        "span",
        "strong",
        "ul",
        "sup",
        "sub",
        "table",
        "td",
        "th",
        "tr",
        "tbody",
        "thead",
        "tfooter",
        "details",
        "summary",
        # 'title',
    ]

    MD_ALLOWED_ATTRIBUTES = {
        "a": ["href", "title", "rel", "data-tooltip"],
        "abbr": ["title"],
        "acronym": ["title"],
        "div": ["class"],
        "img": ["alt", "src", "title"],
    }

    MD_ALLOWED_STYLES = []

    # Extracted from https://github.github.com/gfm/#disallowed-raw-html-extension-
    MD_FILTERED_TAGS = [
        "title",
        "textarea",
        "style",
        "xmp",
        "iframe",
        "noembed",
        "noframes",
        "script",
        "plaintext",
    ]

    MD_ALLOWED_PROTOCOLS = ["http", "https", "ftp", "ftps"]

    # Tags constraints
    TAG_MIN_LENGTH = 3
    TAG_MAX_LENGTH = 96

    # Optionnal license groups used for a select input group widget
    # in admin dataset edit view.
    # A list of tuples, each tuple describing a group with its title and
    # a list of licenses associated. Translations are not supported.
    # Example:
    # LICENSE_GROUPS = [
    #     ("Autorités administratives", [
    #         {"value": "lov2", "recommended": True, "description": "Recommandée", "code": "etalab-2.0"},
    #         {"value": "notspecified", "description": "Le Code des relations entre le public et l’administration ne s’applique pas"}]),
    #     ("Tous producteurs", [
    #         {"value": "lov2", "recommended": True, "description": "Recommandée"},
    #         {"value": "cc-by", "code": "CC-BY"},
    #         {"value": "notspecified"}])
    # ]
    LICENSE_GROUPS = None

    # Cache duration for templates.
    TEMPLATE_CACHE_DURATION = 5  # Minutes.

    DELAY_BEFORE_REMINDER_NOTIFICATION = 30  # Days

    HARVEST_ENABLE_MANUAL_RUN = False

    HARVEST_PREVIEW_MAX_ITEMS = 20

    # Development setting to allow minimizing the number of harvested items
    HARVEST_MAX_ITEMS = None

    # Harvesters are scheduled at midnight by default
    HARVEST_DEFAULT_SCHEDULE = "0 0 * * *"

    # The number of days of harvest jobs to keep (ie. number of days of history kept)
    HARVEST_JOBS_RETENTION_DAYS = 365

    # The number of days since last harvesting date when a missing dataset is archived
    HARVEST_AUTOARCHIVE_GRACE_DAYS = 7

    HARVEST_VALIDATION_CONTACT_FORM = None

    HARVEST_MAX_CATALOG_SIZE_IN_MONGO = None  # Defaults to the size of a MongoDB document
    HARVEST_GRAPHS_S3_BUCKET = None  # If the catalog is bigger than `HARVEST_MAX_CATALOG_SIZE_IN_MONGO` store the graph inside S3 instead of MongoDB
    HARVEST_GRAPHS_S3_FILENAME_PREFIX = ""  # Useful to store the graphs inside a subfolder of the bucket. For example by setting `HARVEST_GRAPHS_S3_FILENAME_PREFIX = 'graphs/'`

    # S3 connection details
    S3_URL = None
    S3_ACCESS_KEY_ID = None
    S3_SECRET_ACCESS_KEY = None

    # Specific support for hvd (map HVD categories URIs to keywords)
    HVD_SUPPORT = True

    ACTIVATE_TERRITORIES = False
    # The order is important to compute parents/children, smaller first.
    HANDLED_LEVELS = tuple()

    LINKCHECKING_ENABLED = True
    # Resource types ignored by linkchecker
    LINKCHECKING_UNCHECKED_TYPES = ("api",)
    LINKCHECKING_IGNORE_DOMAINS = []
    LINKCHECKING_IGNORE_PATTERNS = ["format=shp"]
    LINKCHECKING_MIN_CACHE_DURATION = 60  # in minutes
    LINKCHECKING_MAX_CACHE_DURATION = 1080  # in minutes (1 week)
    LINKCHECKING_UNAVAILABLE_THRESHOLD = 100
    LINKCHECKING_DEFAULT_LINKCHECKER = "no_check"

    # Ignore some endpoint from API tracking
    # By default ignore the 3 most called APIs
    TRACKING_BLACKLIST = [
        "api.notifications",
        "api.check_dataset_resource",
        "api.avatar",
    ]

    DELETE_ME = True

    # Optimize uploaded images
    FS_IMAGES_OPTIMIZE = True

    # Default resources extensions whitelist

    ALLOWED_ARCHIVE_EXTENSIONS = [
        # Archives
        "tar",
        "gz",
        "tgz",
        "rar",
        "zip",
        "7z",
        "xz",
        "bz2",
    ]

    ALLOWED_RESOURCES_EXTENSIONS = ALLOWED_ARCHIVE_EXTENSIONS + [
        # Base
        "csv",
        "txt",
        "json",
        "pdf",
        "xml",
        "rtf",
        "xsd",
        # OpenOffice
        "ods",
        "odt",
        "odp",
        "odg",
        # Microsoft Office
        "xls",
        "xlsx",
        "doc",
        "docx",
        "pps",
        "ppt",
        # Images
        "jpeg",
        "jpg",
        "jpe",
        "gif",
        "png",
        "dwg",
        "svg",
        "tiff",
        "ecw",
        "svgz",
        "jp2",
        # Geo
        "shp",
        "kml",
        "kmz",
        "gpx",
        "shx",
        "ovr",
        "geojson",
        "gpkg",
        # Meteorology
        "grib2",
        # RDF
        "rdf",
        "ttl",
        "n3",
        # Misc
        "dbf",
        "prj",
        "sql",
        "sqlite",
        "db",
        "epub",
        "sbn",
        "sbx",
        "cpg",
        "lyr",
        "owl",
        "dxf",
        "ics",
        "ssim",
        "other",
    ]

    ALLOWED_RESOURCES_MIMES = [
        "text/csv",
        "application/json",
        "application/pdf",
        "application/msword",
        "application/vnd",
        "application/vnd.geo+json",
        "application/zip",
        "application/x-tar",
        "application/x-bzip",
        "application/x-bzip2",
        "application/x-7z-compressed",
        "application/x-rar-compressed",
        "application/epub+zip",
        "application/rtf",
        "application/rdf+xml",
        "application/xml",
        "application/xhtml+xml",
        "image/bmp",
        "image/jpeg",
        "image/png",
        "image/svg+xml",
        "text/html",
        "text/calendar",
        "text/plain",
        "text/xml",
        "text/turtle",
    ]

    # How much time upload chunks are kept before cleanup
    UPLOAD_MAX_RETENTION = 24 * HOUR

    # Avatar providers parameters
    # Overrides themes and default parameters
    # if set to anything else than `None`
    ###########################################################################
    # avatar provider used to render user avatars
    AVATAR_PROVIDER = None
    # Number of blocks used by the internal provider
    AVATAR_INTERNAL_SIZE = None
    # List of foreground colors used by the internal provider
    AVATAR_INTERNAL_FOREGROUND = None
    # Background color used by the internal provider
    AVATAR_INTERNAL_BACKGROUND = None
    # Padding (in percent) used by the internal provider
    AVATAR_INTERNAL_PADDING = None
    # Skin (set) used by the robohash provider
    AVATAR_ROBOHASH_SKIN = None
    # The background used by the robohash provider.
    AVATAR_ROBOHASH_BACKGROUND = None

    # Post settings
    ###########################################################################
    # Discussions on posts are disabled by default
    POST_DISCUSSIONS_ENABLED = False
    # Default pagination size on listing
    POST_DEFAULT_PAGINATION = 20

    # Organization settings
    ###########################################################################
    # The business identification format to use for validation
    ORG_BID_FORMAT = "siret"

    # Preview settings
    ###########################################################################
    # Preview mode can be either `iframe` or `page` or `None`
    PREVIEW_MODE = "iframe"

    # URLs validation settings
    ###########################################################################
    # Whether or not to allow private URLs (private IPs...) submission
    URLS_ALLOW_PRIVATE = False
    # Whether or not to allow local URLs (localhost...) submission.
    URLS_ALLOW_LOCAL = False
    # Whether or not to allow credentials in URLs submission.
    URLS_ALLOW_CREDENTIALS = True
    # List of allowed URL schemes.
    URLS_ALLOWED_SCHEMES = ("http", "https", "ftp", "ftps")
    # List of allowed TLDs.
    URLS_ALLOWED_TLDS = tld_set

    # Flask-CDN options
    # See: https://github.com/libwilliam/flask-cdn#flask-cdn-options
    # If this value is defined, toggle static assets on external domain
    CDN_DOMAIN = None
    # Don't check timestamp on assets (and avoid error on missing assets)
    CDN_TIMESTAMP = False

    # Export CSVs of model objects as resources of a dataset
    ########################################################
    EXPORT_CSV_MODELS = (
        "dataset",
        "resource",
        "discussion",
        "organization",
        "reuse",
        "tag",
        "harvest",
    )
    EXPORT_CSV_DATASET_ID = None

    # Autocomplete parameters
    #########################
    SEARCH_AUTOCOMPLETE_ENABLED = True
    SEARCH_AUTOCOMPLETE_DEBOUNCE = 200  # in ms

    # Archive parameters
    ####################
    ARCHIVE_COMMENT_USER_ID = None
    ARCHIVE_COMMENT_TITLE = _("This dataset has been archived")

    # Schemas parameters
    ####################
    SCHEMA_CATALOG_URL = None

    API_DOC_EXTERNAL_LINK = (
        "#"
    )

    # Read Only Mode
    ####################
    # This mode can be used to mitigate a spam attack for example.
    # It disables the methods listed in the following block list.
    READ_ONLY_MODE = False
    METHOD_BLOCKLIST = [
        "OrganizationListAPI.post",
        "ReuseListAPI.post",
        "DatasetListAPI.post",
        "CommunityResourcesAPI.post",
        "UploadNewCommunityResources.post",
        "DiscussionAPI.post",
        "DiscussionsAPI.post",
        "SourcesAPI.post",
        "FollowAPI.post",
    ]

    FIXTURE_DATASET_SLUGS = []
    PUBLISH_ON_RESOURCE_EVENTS = False
    RESOURCES_ANALYSER_URI = "http://localhost:8000"
    RESOURCES_ANALYSER_API_KEY = None

    # Datasets quality settings
    ###########################################################################
    QUALITY_DESCRIPTION_LENGTH = 100

    # Spam settings
    ###########################################################################
    SPAM_WORDS = []
    SPAM_ALLOWED_LANGS = []
    SPAM_MINIMUM_STRING_LENGTH_FOR_LANG_CHECK = 30

    # Notification settings
    ###########################################################################
    MATTERMOST_WEBHOOK = None

    # Tabular API Dataservice ID
    ###########################################################################
    TABULAR_API_DATASERVICE_ID = None


class Testing(object):
    """Sane values for testing. Should be applied as override"""

    TESTING = True
    # related to https://github.com/noirbizarre/flask-restplus/commit/93e412789f1ef8d1d2eab837f15535cf79bd144d#diff-68876137696247abc8c123622c73a11f  # noqa
    # this keeps our legacy tests from failing, we should probably fix the tests instead someday
    PROPAGATE_EXCEPTIONS = False
    SEND_MAIL = False
    WTF_CSRF_ENABLED = False
    AUTO_INDEX = False
    CELERY_TASK_ALWAYS_EAGER = True
    CELERY_TASK_EAGER_PROPAGATES = True
    TEST_WITH_PLUGINS = False
    PLUGINS = []
    TEST_WITH_THEME = False
    THEME = "testing"
    CACHE_TYPE = "flask_caching.backends.null"
    CACHE_NO_NULL_WARNING = True
    DEBUG_TOOLBAR = False
    SERVER_NAME = "local.test"
    DEFAULT_LANGUAGE = "fr"
    ACTIVATE_TERRITORIES = False
    LOGGER_HANDLER_POLICY = "never"
    CELERYD_HIJACK_ROOT_LOGGER = False
    URLS_ALLOW_LOCAL = True  # Test server URL is local.test
    URLS_ALLOWED_TLDS = tld_set | set(["test"])
    URLS_ALLOW_PRIVATE = False
    FS_IMAGES_OPTIMIZE = True
    SECURITY_EMAIL_VALIDATOR_ARGS = {
        "check_deliverability": False
    }  # Disables deliverability for email domain name
    PUBLISH_ON_RESOURCE_EVENTS = False


class Debug(Defaults):
    DEBUG = True
    SEND_MAIL = False
    DEBUG_TB_INTERCEPT_REDIRECTS = False
    DEBUG_TB_PANELS = (
        "flask_debugtoolbar.panels.versions.VersionDebugPanel",
        "flask_debugtoolbar.panels.timer.TimerDebugPanel",
        "flask_debugtoolbar.panels.headers.HeaderDebugPanel",
        "flask_debugtoolbar.panels.request_vars.RequestVarsDebugPanel",
        "flask_debugtoolbar.panels.config_vars.ConfigVarsDebugPanel",
        "flask_debugtoolbar.panels.template.TemplateDebugPanel",
        "flask_debugtoolbar.panels.logger.LoggingPanel",
        "flask_debugtoolbar.panels.profiler.ProfilerDebugPanel",
        "flask_mongoengine.panels.MongoDebugPanel",
    )
    CACHE_TYPE = "flask_caching.backends.null"
    CACHE_NO_NULL_WARNING = True
